src/userSurvey.js

```diff
@@ -190,37 +214,89 @@ export async function setSurvey(options) {
-        () => {
-            storageSpace.set("surveyCompleted", true);
-            setPopupToNoPromptPage();
-        },
+        setSurveyComplete,
         { urls: [ (new URL(options.surveyCompletionUrl)).href + "*" ] }
     );
 
-    // Set listeners for cancel and open survey button clicks in the survey request
-    messaging.onMessage.addListener(() => {
-        storageSpace.set("surveyCancelled", true);
-        setPopupToNoPromptPage();
-    }, { type: "webScience.userSurvey.cancelSurvey" });
-    messaging.onMessage.addListener(openSurveyInNewTab, { type: "webScience.userSurvey.openSurvey" });
+    // Listeners for cancel and open survey button click only need to be added once.
+    // They do not need to be added again for subsequent calls to setSurvey.
+    if (listenersRegistered === false) {
+        // Set listeners for cancel and open survey button clicks in the survey request.
+        messaging.onMessage.addListener(() => {
+            storageSpace.set("surveyCancelled", true);
+            setPopupToNoPromptPage();
+        }, { type: "webScience.userSurvey.cancelSurvey" });
+        messaging.onMessage.addListener(openSurveyInNewTab, { type: "webScience.userSurvey.openSurvey" });
+    }
+
+    listenersRegistered = true;
 }
 
 /**
  * Each study participant has a persistent survey ID, generated with
  * the id module. The ID is automatically added as a parameter to
  * the survey URL, enabling researchers to import survey data from an
- * external platform and sync it with Rally data.
+ * external platform and sync it with Rally data. This method returns the
+ * survey ID (generating it if it does not already exist)
  * @returns {string} - The participant's survey ID.
  */
 export async function getSurveyId() {
-    return await storageSpace.get("surveyId");
+    initializeStorage();
+    let surveyId = await storageSpace.get("surveyId");
+    if (surveyId === null) {
+        surveyId = id.generateId();
+        await storageSpace.set("surveyId", surveyId);
+    }
+    return surveyId;
+}
+
+/**
+ * Gets the completion status of the current survey. Can be used if a
+ * subsequent survey depends on the status of the previous survey.
+ * @returns {boolean} - Whether the current survey has been completed.
+ * Returns null if there is no current survey.
+ */
+export async function getSurveyCompletionStatus() {
+    initializeStorage();
+    return await storageSpace.get("surveyCompleted");
+}
+
+/**
+ * Gets the name of the current survey.
+ * @returns {string} - The name of the current survey. Returns null
+ * if there is no current survey.
+ */
+export async function getCurrentSurveyName() {
+    initializeStorage();
+    return await storageSpace.get("currentSurvey");
 }
+
+/**
+ * End the current survey. Should be called before a subsequent survey is started.
+ */
+ export async function endCurrentSurvey() {
+    // If there is an existing survey reminder timeout, clears the timeout.
+    if (reminderTimeoutId !== null) {
```

Comment:
Don't need this check, `clearTimeout` will (by specification) just do nothing with an invalid timeout ID.